### PR TITLE
add gnu getopt

### PR DESCRIPTION
**Pull request by tomkinsc**

### CODE REVIEW

**Comment by tomkinsc** [comment]
@conda-forge/core: Is this a build system issue related to https://github.com/conda-forge/idna-feedstock/pull/5? What is the suggested resolution?

**Comment by jakirkham** [comment]
Yes there was. Should be fixed now. Toggling to restart CIs.

**Comment by tomkinsc** [comment]
Thanks, looks the build passed.

**Comment by tomkinsc** [comment]
Can this be merged in?

File: appveyor.yml [hunk 1]
```diff
@@ -67,8 +67,7 @@ install:
     - cmd: call %CONDA_INSTALL_LOCN%\Scripts\activate.bat
     - cmd: conda config --add channels conda-forge
     - cmd: conda config --set show_channel_urls true
-    - cmd: appveyor-retry conda update --yes --quiet conda
-
+    - cmd: conda update --yes --quiet conda
```
Comment (jakirkham) [inline appveyor.yml]: Please revert this change to the `appveyor.yml` file.

File: recipes/gnu-getopt/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+package: 
+  name: {{ name | lower }}
+  version: {{ version }}
+
+source:
+  fn: {{ name }}-{{ version }}.tar.gz  # [osx]
+  md5: {{ md5 }}  # [osx]
+  url: http://frodo.looijaard.name/system/files/software/getopt/getopt-{{ version }}.tar.gz  # [osx]
+  patches:
+    - Makefile.osxpatch  # [osx]
```
Comment (jakirkham) [inline recipes/gnu-getopt/meta.yaml]: No need for all the selectors since it only builds on macOS.

File: recipes/gnu-getopt/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+build:
+  number: 0
+  skip: True  # [not osx]
+
+requirements:
+  build:
+    - gcc  # [osx]
```
Comment (jakirkham) [inline recipes/gnu-getopt/meta.yaml]: Please replace with `toolchain`.

File: recipes/gnu-getopt/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    - gettext  # [osx]
+  run:
+    - libgcc  # [osx]
```
Comment (jakirkham) [inline recipes/gnu-getopt/meta.yaml]: Please drop this.

File: recipes/gnu-getopt/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    - gettext  # [osx]
+
+test:
+    commands:
+        - '$PREFIX/bin/gnu-getopt -V'  # > /dev/null'
```
Comment (jakirkham) [inline recipes/gnu-getopt/meta.yaml]: Single quotes don't seem like a good idea here given the environment variable.

File: recipes/gnu-getopt/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+about:
+    home: 'http://software.frodo.looijaard.name/getopt/'
+    license: "GPL-2.0"
+    license_family: GPL
+    license_file: COPYING
+    summary: "Command-line option parsing library"
+
+extra:
+  notes: "On Linux systems gnu-getopt is simply called 'getopt', however due to potential collision with the native BSD getopt on OSX, this binary is renamed 'gnu-getopt'."
```
Comment (jakirkham) [inline recipes/gnu-getopt/meta.yaml]: Please wrap this line.

File: recipes/gnu-getopt/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    - libgcc  # [osx]
+    - gettext  # [osx]
+
+test:
+    commands:
+        - '$PREFIX/bin/gnu-getopt -V'  # > /dev/null'
+
+about:
+    home: 'http://software.frodo.looijaard.name/getopt/'
+    license: "GPL-2.0"
```
Comment (jakirkham) [inline recipes/gnu-getopt/meta.yaml]: No need for quotes with either of these.

File: recipes/gnu-getopt/build.sh [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+
+export CFLAGS="-I${PREFIX}/include"
+export CPPFLAGS="-I${PREFIX}/include"
+export LDFLAGS="-L${PREFIX}/lib -L${PREFIX}/include -lintl"
```
Comment (jakirkham) [inline recipes/gnu-getopt/build.sh]: Please make sure the original environment variable's value is included in the assignment.

File: recipes/gnu-getopt/build.sh [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+cd "$SRC_DIR"
```
Comment (jakirkham) [inline recipes/gnu-getopt/build.sh]: This is already done by `conda-build`. So can be dropped.

File: recipes/gnu-getopt/build.sh [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+make
+make install prefix=$PREFIX
+
+# rename to prevent collision with system (BSD) getopt on OSX
+mv $PREFIX/bin/getopt $PREFIX/bin/gnu-getopt
```
Comment (jakirkham) [inline recipes/gnu-getopt/build.sh]: Seems like if this is a concern one can simply not install this package and use the system copy, right? Or am I missing something. 🤔

**Comment by jakirkham** [comment]
So there are a few things that still need to be addressed in the recipe. Please let me know if you have any questions.

File: recipes/gnu-getopt/build.sh (hunk 10 above)
Comment (tomkinsc) [reply recipes/gnu-getopt/build.sh]: The BSD version of getopt is significantly divergent from the GNU version (or rather, vise versa), with different arguments and combatibilities. Some things are written with the BSD version in mind, others with the GNU version. This at leaves it open to alias or symlink to the GNU version, if desired.

File: recipes/gnu-getopt/meta.yaml (hunk 7 above)
Comment (tomkinsc) [reply recipes/gnu-getopt/meta.yaml]: The quotes do not make a difference here, and honestly this is sort of pickiness that has prevented me from adding packages to conda-forge, despite being a core contributor of many packages to bioconda. The more substantive changes in this PR I can address, but two bytes for quotes are really a blocking issue?

File: appveyor.yml (hunk 1 above)
Comment (tomkinsc) [reply appveyor.yml]: I'll update this to the version in the current master.

File: recipes/gnu-getopt/meta.yaml (hunk 3 above)
Comment (tomkinsc) [reply recipes/gnu-getopt/meta.yaml]: Sure, but where is this documented? It is not part of the linter.

File: recipes/gnu-getopt/meta.yaml (hunk 4 above)
Comment (tomkinsc) [reply recipes/gnu-getopt/meta.yaml]: Sure.

**Comment by tomkinsc** [comment]
Thanks for the review. A number of comments in reply.

File: recipes/gnu-getopt/meta.yaml (hunk 5 above)
Comment (tomkinsc) [reply recipes/gnu-getopt/meta.yaml]: The quote variant shouldn't matter since the string should hit the yaml parser first, no? Changing them out I can remove the comment though...

File: recipes/gnu-getopt/meta.yaml (hunk 6 above)
Comment (tomkinsc) [reply recipes/gnu-getopt/meta.yaml]: Sure.

File: recipes/gnu-getopt/build.sh (hunk 8 above)
Comment (tomkinsc) [reply recipes/gnu-getopt/build.sh]: Sure.

File: recipes/gnu-getopt/build.sh (hunk 9 above)
Comment (tomkinsc) [reply recipes/gnu-getopt/build.sh]: Right, thanks.

File: recipes/gnu-getopt/meta.yaml (hunk 2 above)
Comment (tomkinsc) [reply recipes/gnu-getopt/meta.yaml]: True!

**Comment by tomkinsc** [comment]
Any additional reasons why this cannot be merged?
